tokio-executor/src/task/core.rs

```diff
@@ -0,0 +1,158 @@
+use crate::loom::alloc::Track;
+use crate::loom::cell::CausalCell;
+use crate::task::Schedule;
+use crate::task::state::State;
+use crate::task::raw::{self, Vtable};
+use crate::task::waker::waker_ref;
+
+use std::cell::UnsafeCell;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::task::{Context, Poll, Waker};
+
+/// The task cell. Contains the components of the task.
+///
+/// It is critical for `Header` to be the first field as the task structure will
+/// be referenced by both *mut Cell and *mut Header.
+#[repr(C)]
+pub(super) struct Cell<T: Future, S: 'static> {
+    /// Hot task state data
+    pub(super) header: Header<S>,
+
+    /// Either the future or output, depending on the execution stage.
+    pub(super) core: Core<T>,
+
+    /// Cold data
+    pub(super) trailer: Trailer,
+}
+
+/// The core of the task.
+///
+/// Holds the future or output, depending on the stage of execution.
+pub(super) struct Core<T: Future> {
+    stage: Stage<T>,
+}
+
+/// Crate public as this is also needed by the pool.
+#[repr(C)]
+pub(crate) struct Header<S: 'static> {
+    /// Task state
+    pub(super) state: State,
+
+    /// Pointer to the executor owned by the task
+    pub(super) executor: CausalCell<Option<NonNull<S>>>,
+
+    /// Pointer to next task, used for misc task linked lists.
+    pub(crate) queue_next: UnsafeCell<*const Header<S>>,
```

Comment:
Should this be changed to an `Option<NonNull<Header<S>>>` as well, if only for consistency with the `owned` list?